pkg/porter/credentials.go

```diff
@@ -146,6 +151,86 @@ func (o *CredentialShowOptions) Validate(args []string) error {
 	return o.ParseFormat()
 }
 
+// Validate validates the args provided to Porter's credential edit command
+func (o *CredentialEditOptions) Validate(args []string) error {
+	if err := validateCredentialName(args); err != nil {
+		return err
+	}
+	o.Name = args[0]
+	return nil
+}
+
+// EditCredential edits the credentials of the provided name.
+func (p *Porter) EditCredential(opts CredentialEditOptions) error {
+	credSet, err := p.Credentials.Read(opts.Name)
+	if err != nil {
+		return err
+	}
+
+	if len(credSet.Credentials) == 0 {
+		return errors.New("no credentials exist in this set")
+	}
+
+	// use a flattened data structure for the credentials, so when we marshal
+	// the structure to yaml it's easier for the user to edit.
+	type tempStrategy struct {
+		Name        string `yaml:"name"`
+		SourceType  string `yaml:"type"`
+		SourceValue string `yaml:"value"`
+	}
+	type tempCredSet struct {
+		Credentials []tempStrategy `yaml:"credentials"`
+	}
+	tempSet := &tempCredSet{}
+	for _, cs := range credSet.Credentials {
+		tempSet.Credentials = append(tempSet.Credentials, tempStrategy{
+			Name:        cs.Name,
+			SourceType:  cs.Source.Key,
+			SourceValue: cs.Source.Value,
+		})
+	}
+
+	data, err := yaml.Marshal(tempSet)
+	if err != nil {
+		return err
+	}
+
+	editor := &Editor{
+		Data:     []byte(data),
+		Filename: fmt.Sprintf("porter-%s", credSet.Name),
+		Context:  p.Context,
+	}
+
+	output, err := editor.Run()
+	if err != nil {
+		return err
+	}
+
+	err = yaml.Unmarshal(output, tempSet)
+	if err != nil {
+		return err
+	}
+
+	newCredentials := []credentials.CredentialStrategy{}
+	// todo: validate for correct source types
```

Comment:
Right now we don't have a way to dynamically validate, so you will need to compare against a known list, if you look in the credential generator you can find it.

Reply:
Sounds good, I'll use that for the validation.

Reply:
Question around this: should credentials.EditCredentials be responsible for validating credentials, or should that go somewhere upstream, like the credential provider? I was looking at potentially calling Resolve from here to validate the sources, but I'm not 100% sure. Thoughts?

Reply:
Calling Resolve wouldn't do what you are hoping for, since that would actually try to find a value that matches the credential, for example an environment variable named "AZURE_CLIENT_ID", and if one can't be found, it would fail.

Yes, this could be a new function on the CredentialProvider interface, `Validate`, and then implemented on CredentialStorage. Then for now we can check against the set of constants that we know are valid source keys.